docs/RFCS/20210527_changefeed_http_sink.md

```diff
@@ -0,0 +1,202 @@
+struct might look something like this:
+
+```
+type httpSink struct {
+  u sinkURL
+  client *httputil.Client
+	mu struct {
+		syncutil.Mutex
+		inflight int64
+		flushCh chan struct{}
+	}
+}
+```
+
+## Concurrency
+
+Each `POST` to the HTTP endpoint is executed asynchronously through calls to 
+`EmitRow()` and `EmitResolvedTimestamp()`. We use a channel here, `flushCh`,
+which we will send the http response object once it is received, or the error
+(if one occurs). We use a channel as opposed to something like a `WaitGroup` to
+receive a response (which may be needed for ack functionality later on) or an 
+error.
+
+```
+func (s *httpSink) EmitRow (...) error {
+  ...
+  res, err := s.client.Post(...)
+  s.mu.Lock()
+  s.mu.inflight++
+  s.mu.Unlock()
+  go func() {
+    s.mu.flushCh <- struct {
+      res *http.Response
+      err error
+    } {
+      res: res,
+      err: err,
+    }
+  }
+  ...
+  return err
+}
+```
+
+On the `Flush()` side, the the `mu` field can act as a semaphore, and allow
+reading from the channel until the sink has been flushed.
+
+```
+func (s *httpSink) Flush (...) error {
+  ...
+  s.mu.Lock()
+  defer s.mu.Unlock()
+  for s.mu.inflight > 0 {
+    select {
+      case ctx.Done():
+        return ctx.Err()
+      case flushed := <- s.mu.flushCh:
+        s.mu.inflight--
+        // can add additional ack logic here with flushed.res if needed
+        log.Infof("Sink returned response: %v", flushed.res)
+  }
+  ...
+}
+```
+
+## Configurability
+
+The HTTP sink should be configurable through a few different options and sink
+URL query params, a few of which are already supported by other sinks:
+
+* `ca_cert` - supply a custom CA cert for server verification
+* `insecure_tls_skip_verify` - used to disable client-side verification of CA
+  certs (corresponds to Go's 
+  [`tls.Config.InsecureSkipVerify`](https://golang.org/src/crypto/tls/common.go#L624) 
+  flag)
+
+These TLS-related query params should only be enabled if the provided sink URL
+is prefixed with `https`, throwing an error otherwise.
+
+Some HTTP-based options should also be added for `CREATE CHANGEFEED` statement:
+
+* `http_auth_header` - a user-supplied `Authorization` header for either basic 
+  base64-encoded auth or an OAuth token. As this contains credentials, it
+  should be redacted when appearing in the `description` column of `SHOW JOBS`/
+  `SHOW CHANGEFEED JOBS` etc.
+* `http_client_timeout` - specifies a timeout (in seconds) for the client to 
+  wait for each request to complete (corresponds to Go's 
+  [`http.client.Timeout`](https://golang.org/src/net/http/client.go#L105)) 
+  If no option is provided, the default timeout is 3 seconds.
+
+Altogether, the command to create a changefeed to an HTTP sink would look 
+something like this:
+
+```
+$ root@:26257/defaultdb> CREATE CHANGEFEED FOR foo INTO 
+                      -> INTO 'https://fake-endpoint:8080?ca_cert=Zm9v...'
```

Comment:
I think this example is actually good evidence that passing configuration as query parameters is perhaps UX we should avoid.  How does the user intuit which configuration should be a query parameter and which should be an option?